dashboard/test/controllers/programming_methods_controller_test.rb

```diff
@@ -0,0 +1,69 @@
+require 'test_helper'
+
+class ProgrammingMethodsControllerTest < ActionController::TestCase
+  include Devise::Test::ControllerHelpers
+
+  setup do
+    File.stubs(:write)
+    FileUtils.stubs(:mkdir_p)
+    Rails.application.config.stubs(:levelbuilder_mode).returns true
+    @levelbuilder = create :levelbuilder
+    @programming_class = create :programming_class
+  end
+
+  test 'can update programming method from params' do
+    sign_in @levelbuilder
+    File.expects(:write).once
+
+    programming_method = create :programming_method, programming_class: @programming_class
+
+    post :update, params: {
+      id: programming_method.id,
+      key: programming_method.key,
+      name: 'new name',
+      content: 'a longer description of the code',
+      syntax: 'new Class()',
+      examples: [{name: 'example 1', embed_app_with_code_height: '300px'}]
+    }
+    assert_response :ok
+    programming_method.reload
+
+    assert_equal 'new name', programming_method.name
+    assert_equal 'a longer description of the code', programming_method.content
+    assert_equal 'new Class()', programming_method.syntax
+    assert_equal [{name: 'example 1', embed_app_with_code_height: '300px'}].to_json, programming_method.examples
+  end
+
+  test 'data is passed down to edit page' do
+    sign_in @levelbuilder
+
+    programming_method = create :programming_method, programming_class: @programming_class
+
+    get :edit, params: {id: programming_method.id}
+    assert_response :ok
+
+    edit_data = JSON.parse(css_select('script[data-programmingmethod]').first.attribute('data-programmingmethod').to_s)
+    assert_equal programming_method.key, edit_data['key']
+    assert_equal programming_method.name, edit_data['name']
+  end
+
+  class ProgrammingMethodsControllerAccessTests < ActionController::TestCase
```

Comment:
Interesting, haven't seen this pattern before. Is this something we do commonly in levelbuilder controllers? Kind of a nice delineation

Reply:
I started using this because some of our other controller tests (looking at you scripts_controller_test) are really slow due to shared setup code. Creating a separate "class" here redefines that setup code, so it can prevent setup creep. I ended up liking this patters of delineating separate but related tests so I kept doing it. Definitely a style preference on this particular test file as there aren't enough tests or setup for there to be performance gains here